Loadtest stableapis

krishna1s opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [x] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/21314

Commit:
publishing stable api version 2022-11-01 for load test service

Commit:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into loadtest_stableapis

Commit:
prettier run

krishna1s commented:
@mikekistler Hi Mike, we are releasing the first stable version of our data plane APIs as our service is going into GA soon. Please review and let us know your feedback.

Commit:
spell check

Commit:
changing swagger back to 2.0

krishna1s committed:
adding description

krishna1s committed:
prettier changes

krishna1s committed:
renaming opertation  id

krishna1s committed:
adding x-ms-enum extension

krishna1s committed:
validation fixes

Commit:
renaming examples

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+          {
+            "in": "query",
+            "name": "testRunId",
+            "description": "Required testRunId, if testId field is not provided",
```
Is there no way to simply get a list of appComponents? If not, why not?

krishna1s replied:
added more context on above comment

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+        "tags": [
+          "AppComponent"
+        ],
+        "summary": "Get app component for a test or a test run by unique its name.",
```
I think this is a copy/paste error since `name` is not a parameter to this operation.

krishna1s replied:
operation was swapped, changed now.

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+    "version": "2022-11-01"
+  },
+  "paths": {
+    "/appcomponents/{name}": {
```
Should use kebab-case in paths.
```suggestion
    "/app-components/{name}": {
```

krishna1s replied:
changed

mikekistler reviewed (changes requested):
These are just my comments on the `appcomponents` part of the API. I'll continue reviewing and provide feedback on the rest of the API but wanted to get the discussion started on these.

Commit:
review changes

Commit:
adding x-ms-parameter-location

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+  },
+  "paths": {
+    "/appcomponents/{name}": {
+      "patch": {
```
@mikekistler I am adding more context to this operation. We store customer azure resource details as configuration so that we can pull azure resource metrics, along with our load test metrics and this helps customers to compare the load test effect on their application resources. AppComponent entity stores a list of azure resource details (like resource-id, type, name, etc.). This is required to pull the metrics info from the azure monitor. This app component can be attached with a test or test run. So for this operation, we take AppComponent ( which is basically a map of azure resources) & test, test run id to associate with, and combine all this we call as AppcomponentMap. 
Considering this flow, let us know if the request body and response are correct or not. Let us know if there is a better way to design this API as well.

mikekistler replied:
Let me explain my comment a bit more. The convention for REST APIs is that when a path ends in a static path segment, like "app-components", that last segment is the name of a "collection" of resources. It is always plural, because it is a collection, but the resources of the collection are individual instances of that resource, so in this case "app-component".

It seems that you are trying to use this endpoint to define a "collection of collections" -- this is NOT conventional in REST.

And it is even more confusing because the description of the operation says it associates an "app component (Azure resource)" (singular) "to a test or test run".

One possible solution is to simply embed the app components map directly into the Test and TestRuns schemas. Just make it a property of these other resource models. This would simplify your API by eliminating all the /app-component operations. 

Another possibility is to make "app-components" a sub-path of /loadtests/{testId} and also /testruns/{testRunId} with get and patch methods (and possibly a delete method -- not sure that's needed).

krishna1s replied:
We have implemented the suggested approach

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+          "200": {
+            "description": "The requested app components",
+            "schema": {
+              "$ref": "#/definitions/AppComponentsMap"
```
added more context above

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "/serverMetricsConfig": {
```
As with app-components, I think serverMetricsConfig (which should be kebab-case "server-metrics-config" when in the path) should either be simply embedded into the Test and TestRuns schemas or a sub-path of /loadtests/{testId} and /testruns/{testRunId}.

krishna1s replied:
changed to kebab case

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "/loadtests/{testId}": {
```
Recommend changing this path to simply "/tests". This will be consistent with 1) the response model name (after it is fixed to remove "model") 2) the external docs, and 3) the "/testruns" path, which is not "/loadtestruns".

krishna1s replied:
changed

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "/loadtests/sortAndFilter": {
```
This is a standard list operation on the loadtests collection so it should be a GET on "/loadtests"

krishna1s replied:
changed

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+              "USER_PROPERTIES",
+              "ADDITIONAL_ARTIFACTS"
+            ],
+            "x-ms-enum": {
```
Please add descriptions for all values in the x-ms-enum extension. This will benefit the generated REST API docs.

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "/testruns/{testRunId}": {
```
Static path segments should be kebab-case: /testruns -> /test-runs

krishna1s replied:
changed

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "/testruns/sortAndFilter": {
```
This is a standard list operation on the testruns collection so it should be a GET on "/testruns"

krishna1s replied:
changed

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "/testRuns/{testRunId}/metricDefinitions": {
```
Note inconsistent casing here: "testRuns" vs "testruns". This reinforces the earlier comment that these should be kebab-case: /testRuns -> /test-runs

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "CertificateMetadata20221101": {
```
Please remove the "20221101" suffix from everything.

krishna1s replied:
removed

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "DefaultServerMetricsConfigListModel": {
```
Please drop the "Model" suffix from all your schema names.

krishna1s replied:
removed

mikekistler reviewed (changes requested):
Here is another batch of comments. I know some of these are not minor changes but I think this will make your API much easier for developers to understand and use it.

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+        }
+      }
+    },
+    "LocalizableString": {
```
Our API doesn't support yet localization, we try to keep the same schema as azure monitor REST APIs( as in the future we might move to Azure monitor for our Load test metrics). Should we rename the model?

msyyc commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2958 @@
+            "in": "formData",
+            "name": "file",
+            "required": true,
+            "type": "file"
```
lack of "description"

krishna1s replied:
added

dw511214992 reviewed (commented):
It's strongly to follow the [doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/747/Service-Onboard-DPG-with-Swagger-CI-Pipeline) to enable sdk automation. Then you will be able to preview the generated sdk and apiview in PR comment. 
I create another PR based on your PR and you can refer to it as example: https://github.com/Azure/azure-rest-api-specs/pull/21177 (Please ignore the failure of .Net SDK because .Net SDK Codegen has some issues now.)

dw511214992 commented on specification/loadtestservice/data-plane/readme.md:
````diff
@@ -23,7 +23,16 @@ These are the global settings for the loadtestservice.
 ```yaml
 openapi-type: data-plane
-tag: package-2022-06-01-preview
+tag: package-2022-11-01
````
```suggestion
tag: package-2022-11-01
security: AADToken
security-scopes: https://loadtest.azure-dev.com/.default
title: LoadTestingClient
```

mikekistler replied:
I would prefer that the security/security-scopes be specified in the REST API definition and skip them here.

dw511214992 replied:
@mikekistler Yes, if it's defined in swagger, there is no need to add it to readme.md

Harshan01 replied:
Removed as suggested

Commit:
review fixes

Commit:
removing required field

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+      },
+      "description": "Azure Active Directory OAuth2 Flow"
+    }
+  },
```
added

krishna1s committed:
adding x-ms-enum values

krishna1s committed:
prettier fix

krishna1s committed:
spell fixes

krishna1s committed:
prettier fix

Commit:
adding new metrics and app component apis

Commit:
updating examples

Commit:
removing list resource type api

Commit:
changing to kebab case

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,2962 @@
+          }
+        }
+      },
+      "delete": {
```
We will do this later in the next release

Commit:
changing x-ms-parameter-location value for sdk

Commit:
updating example

Commit:
removing resourceId

Commit:
removing resource Id from examples

Commit:
adding missing desc

Commit:
removing read only properties from examples

Commit:
spell fix

Commit:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into loadtest_stableapis

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+          "type": "string"
+        },
+        "fileId": {
+          "description": "File unique identifier.",
```
Presumably this is the same value as the "fileId" parameter used to create the file. That parameter is described as "Unique name for test file". I think you should use the same description here. I realize that there is potential confusion with "filename" below, but I think it is important to keep descriptions of inputs and outputs consistent.

As an aside, the service rather consistently describes parameters and properties named XXXId as the "name" of XXX. That too is a little confusing but it is so consistent across the service that I think it is best to stick with that.

krishna1s replied:
description is changed

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+            "type": "string"
+          },
+          {
+            "in": "query",
+            "name": "lastUpdatedStartTime",
+            "description": "Start DateTime(ISO 8601 literal format) of the last updated time range to filter tests.",
+            "type": "string",
+            "format": "date-time"
+          },
+          {
+            "in": "query",
+            "name": "lastUpdatedEndTime",
+            "description": "End DateTime(ISO 8601 literal format) of the last updated time range to filter tests.",
+            "type": "string",
+            "format": "date-time"
+          },
```
With the exception of these two parameters, the service uses "lastModified" rather than "lastUpdated" in parameter/property names. For consistency I recommend you change these to "lastModified".

krishna1s replied:
used as lastModified

mikekistler reviewed (changes requested):
I gave this one more good scrub. There is nothing serious here but fixing these will make for better generated SDKs and I hope an easier review the SDK Architecture board.

weidongxu-microsoft commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+    }
+  },
+  "definitions": {
+    "AzureResourceMetadata": {
```
Seems this indeed refers to ARM resources. Though curious what effect does this ARM resources take in the test run.

krishna1s replied:
@weidongxu-microsoft  we pull out resource specific metrics to compare with load test metrics for customers to compare the impact of a load test run on their application. There is no effect on either load test or azure resource, it just need azure resource info for pulling the metrics

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+            "minLength": 2,
+            "pattern": "^[a-z0-9_-]*$"
+          },
+          {
+            "in": "query",
+            "name": "fileType",
+            "description": "File type",
+            "type": "string",
+            "enum": [
+              "JMX_FILE",
+              "USER_PROPERTIES",
+              "ADDITIONAL_ARTIFACTS"
+            ],
+            "x-ms-enum": {
+              "name": "fileType",
+              "modelAsString": true,
+              "values": [
+                {
+                  "value": "JMX_FILE",
+                  "description": "If file is jmx script"
+                },
+                {
+                  "value": "USER_PROPERTIES",
+                  "description": "If file is user properties"
+                },
+                {
+                  "value": "ADDITIONAL_ARTIFACTS",
+                  "description": "If file is not any of other supported type"
+                }
+              ]
+            }
```
@weidongxu-microsoft if it is not provided we are deriving from the file extension, let us know if we should make it required

weidongxu-microsoft replied:
Sure, current approach is good.

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+            "minLength": 2,
+            "pattern": "^[a-z0-9_-]*$"
+          },
+          {
+            "in": "query",
+            "name": "aggregation",
+            "description": "The list of aggregation types (comma separated) to retrieve.",
+            "type": "string"
+          },
```
@weidongxu-microsoft  aggregation values are known to users when they pull the metric definitions, we have kept this similar to Azure Rest Monitor APIs, - https://learn.microsoft.com/en-us/rest/api/monitor/metrics/list?tabs=HTTP#uri-parameters

let us know if we should change

weidongxu-microsoft replied:
If you are sure user knows what value to input, current is fine. If user might not know, maybe at least mention it in description.

If I am familiar with monitor, I know Average,Count,Minimum,Maximum,Total. But not sure about your user.

krishna1s replied:
we have provided the enum values, to see the metrics supported aggregation, they can pull the metric definition and see the supported aggregation

weidongxu-microsoft replied:
You mean via "/test-runs/{testRunId}/metric-definitions" API?

Sure, look good to me.

krishna1s replied:
yes

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        }
+      }
+    },
+    "/test-runs/{testRunId}/server-metrics-config": {
```
it is map of configuration, will change it like per app-components => server-metric-configs

krishna1s replied:
changed

Commit:
review fixes

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        "parameters": [
+          {
+            "in": "query",
+            "name": "orderBy",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+          },
+          {
+            "in": "query",
+            "name": "maxPageSize",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        "parameters": [
+          {
+            "in": "query",
+            "name": "orderBy",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+            ]
+          }
+        },
+        "expireTime": {
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      "type": "object",
+      "properties": {
+        "value": {
+          "description": "List of Resources",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      "readOnly": true
+    },
+    "TestResourceList": {
+      "description": "List of Resources",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+          "Test"
+        ],
+        "summary": "Get associated app component (collection of azure resources) for the given test.",
+        "operationId": "Test_GetAppComponent",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+            "minLength": 2,
+            "pattern": "^[a-z0-9_-]*$"
+          },
+          {
+            "in": "query",
+            "name": "oldTestRunId",
+            "description": "Existing test run Id that should be rerun.",
+            "type": "string"
+          },
```
added more explanation

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        }
+      }
+    },
+    "InputTestArtifacts": {
```
changed, created separate for test & test run

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      "description": "The input artifacts for the test.",
+      "type": "object",
+      "properties": {
+        "configUrl": {
+          "$ref": "#/definitions/FileUrl"
+        },
+        "testScriptUrl": {
+          "$ref": "#/definitions/FileUrl"
+        },
+        "userPropUrl": {
+          "$ref": "#/definitions/FileUrl"
+        },
+        "inputArtifactsZipFileurl": {
+          "$ref": "#/definitions/FileUrl"
+        },
+        "additionalUrls": {
+          "description": "Additional supported files for the test run",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/FileUrl"
+          },
+          "readOnly": true
+        }
+      },
```
changed the FileUrl suffix to FileInfo

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+          "description": "The number of engine instances to execute load test. Supported values are in range of 1-45. Required for creating a new test.",
+          "type": "integer"
+        },
+        "splitAllCSVs": {
+          "description": "Whether all the input CSV files should be split evenly across all engines.",
+          "type": "boolean"
+        },
```
added default value and more explanation

Commit:
mode renaming

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+          "TestRun"
+        ],
+        "summary": "Lists the metric namespaces for a load test run.",
+        "operationId": "TestRun_MetricNamespaces",
```
changed the operation id as suggested

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        }
+      }
+    },
+    "TestAppComponent": {
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        "tags": [
+          "Test"
+        ],
+        "summary": "Configure server metrics to a test",
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+        "userPropUrl": {
+          "$ref": "#/definitions/FileUrl"
+        },
+        "inputArtifactsZipFileurl": {
```
changed with FileInfo suffix

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      },
+      "readOnly": true
+    },
+    "LoadTestConfig": {
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+          "description": "Azure resource id.",
+          "type": "string"
+        },
+        "metricnamespace": {
```
changed

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      "description": "Environment variable or secret",
+      "type": "object",
+      "properties": {
+        "value": {
+          "description": "The value of the secret for the respective type",
+          "type": "string"
+        },
```
changed to Secret

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      "description": "Represents collection of metric definitions.",
+      "type": "object",
+      "properties": {
+        "value": {
```
added as required

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+      "description": "Collection of file urls.",
+      "type": "object",
+      "properties": {
+        "value": {
```
added

Commit:
properties changes in examples

Commit:
renaming

Commit:
Update loadtestservice.json

Commit:
renaming properties

Commit:
renaming property

johanste commented on specification/loadtestservice/data-plane/readme.md:
```diff
@@ -66,7 +77,7 @@ namespace: com.azure.developer.loadtesting
 output-folder: $(azure-sdk-for-java-folder)/sdk/loadtestservice/azure-developer-loadtesting
 service-versions:
-- 2022-06-01-preview
+- 2022-11-01
```
Why are there still transforms? It seems like all of these transforms can be incorporated into the spec itself?

Harshan01 replied:
We have removed the Java section based on previous discussions. The autorest options would now be present in the dedicated SDK repos only

Commit:
desc changes

krishna1s commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3315 @@
+            "description": "Start DateTime(ISO 8601 literal format) of test-run execution time filter range.",
+            "type": "string",
+            "format": "date-time"
+          },
```
description reversed, thanks for highlighting

Commit:
Merge branch 'main' into loadtest_stableapis

Harshan01 committed:
Add readme SDK changes (#21315)

Co-authored-by: Harshan B <harshanb@microsoft.com>

Harshan01 commented on specification/loadtestservice/data-plane/readme.md:
````diff
@@ -44,14 +53,16 @@ input-file:
 These settings apply only when `--java` is specified on the command line.
 Please also specify `--azure-sdk-for-java-folder=<path to the root directory of your azure-sdk-for-java clone>`.
 
 ```yaml $(java)
````
Done

Harshan01 commented on specification/loadtestservice/data-plane/readme.md:
````diff
@@ -44,14 +53,16 @@ input-file:
   - Microsoft.LoadTestService/preview/2021-07-01-preview/loadtestservice.json
 ```
 
+
````
Added suggested changes

Harshan01 replied:
@dw511214992 With language specific options removed and added to respective repos, all languages would now use default autorest settings right? What is the recommended approach to specify language-specific options in this repo?

Commit:
review changes

1. adding file with application octet stream
2.taking fileName as param instead of fileId
3. adding new api for metrics dimension values
4. simplifying metrics api

Commit:
model validation fixes

Commit:
spell fixes

Commit:
updating desc for search and orderby

Commit:
desc changes

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3258 @@
+            "schema": {
+              "$ref": "#/definitions/TestRun"
+            }
+          },
```
This operation should return a 201 status code in the case where the test run was created.

krishna1s replied:
added a 201 response code

mikekistler commented on specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json:
```diff
@@ -0,0 +1,3258 @@
+        }
+      }
+    },
+    "Metric": {
```
This schema is the response body for "TestRun_GetMetrics" so I think the name should be plural to match the operation name.
```suggestion
    "Metrics": {
```

krishna1s replied:
changed

mikekistler reviewed (approved):
This looks very good! 👍 

I left a few minor comments that I expect are easy fixes so I'll approve now.

Commit:
adding 201 response

krishna1s commented:
@mikekistler do we need any further approval on this PR or can we merge this?

Commit:
adding 201 status code

mikekistler commented:
@jhendrixMSFT I think this PR is good to merge. Will you do the honors please?

krishna1s commented:
/azp run
